# <center>DS4E-S21: Homework 2</center> 
    
    

This homework is due on **Monday**, **March 8**, by **9a ET**. It is worth 34 points (one point per sub-question). Late homeworks, even by one minute, will be graded down with no exceptions. The course academic integrity policy applies to every homework, including this one. 

# Instructions

Please complete your answers in the cells provided and then submit this homework according to the same process as Homework 1. 

- For each **short answer** question, simply type your answer using "Markdown" formatting in the cell below each question. For all questions, you may provide brief answers. There is no need for complete sentences.
- For each **coding** question, simply type and run your code in the cell provided. 
- For questions that require **both** code and a short answer, please use the first answer cell(s) for your code and then add an additional one to provide your answer in Markdown. 

Note: we will only provide one cell for all questions, but you can use as many as you need for your answers -- though for this homework it should rarely, if ever, be more than two total, including Markdown cells. Be sure also to **comment your code** in each code cell, or at least every 2-3 lines if you have more than that per cell (unlikely for this assignment!).

Some questions refer to **external articles & resources**. You can find these by clicking the links provided. Some questions also refer to **external data**. You can find this data in the Homework 2 JupyterHub folder (the same folder where you found this homework).  

Finally, be sure to save your work as you go!

# Question 1: Measuring trust in government

The below questions refer to the data and research described in this article, "[Public Trust in Government 1958-2019](https://www.pewresearch.org/politics/2019/04/11/public-trust-in-government-1958-2019/)" from the Pew Research Center.

**(a) How is "trust in government" *conceptualized* according to the first paragraph and visualization in the article?**

Americans believing the government will do the right thing almost always or most of the time.

**(b) What is one strength of this conceptualization? Briefly explain your answer.**

This conceptualization is specific to Americans. This means that the focus on the research is the American government only and we get viewpoints of people judging the same government and the same actions/policies. 

**(c) What is one weakness of this conceptualization? Briefly explain your answer.**

The "right thing" is subjective and different people have different meanings of it. People have different political opinions, so while someone might think of a government's action as right, another person might not.

**(d) The article itself is not clear about how "trust in government" is *operationalized* (which we as data scientists know is not a great practice!). But, operationalization is "trust in government" is explained in a longer report accompanying the article, which you can find [here](https://www.pewresearch.org/politics/wp-content/uploads/sites/4/2017/05/05-03-17-Trust-release.pdf). Based on this report, how is "trust in government" operationalized?** 

*Note: You don't need to write out every detail, just a short summary of what was done is fine!*

A national survey taken April 5-11,2017, surveying 1,501 adults (18 years or older) from all 50 states and DC. Of the people interviewed (in either English or Spanish), 375 were through landline telephone, 1126 through cellphone, including 698 without a landline telephone. Landline responders were selected by randomly asking for the youngest person at home, while 18+ year olds cellphone responders were selected. They were asked about their level of frustration with the federal government, confidence in US's future, and trust in government to do the right thing always/ most of the time. Their political affiliation, gender, age, educational level where all weighted with data from the census.

**(e) What is one strength of this operationalization? Briefly explain your answer.**

This operationalization is very specific in what information they are gathering and how they are gathering it. It tells us about how many people were in the sample and how it was collected (through landline or cellphone) and what questions were asked.

**(f) What is one weakness of this operationalization? Briefly explain your answer.**

This sample does not take the opinions of people who don't speak English or Spanish into account. This introduces errors of exclusion, because America is a diverse country with people who speaks many different languages, and while they might still have strong opinions about the government, they are not considered in this study.

# Question 2: Sequences of data

For this question we will use some example data from the [original article](https://www.pewresearch.org/politics/2019/04/11/public-trust-in-government-1958-2019/) from Question 1. This time, we will consider the data behind the **second** visualization, which you can see by clicking on the "data" tab at the top of the visualization titled "Trust in govt higher among members of party that controls presidency." 

*Note: If you can't find the data, don't worry! We're just sharing it in case you're interested! We will provide you with all the data you need in this homework.*

For this question, we will only consider the most recent five observations, which reflect the five most recent studies Pew has conducted on this topic. Specifically, suppose we have the following mini dataset:

* Percentage of **Republicans** who say they trust the government in Washington always or most of the time: 11, 11, 28, 22, 21
* Percentage of **Democrats** who say they trust the government in Washington always or most of the time: 17, 26, 15, 15, 4

Each number refers to (approx.) each year between 2014 through 2019. So, for example, in 2014, 11% of Republicans said they trust the government in Washington always or most of the time, and 17% of Democrats said they trust the government in Washington always or most of the time. Note the researchers also seem to have skipped 2018. Welcome to the wide, wonderful world of messy data!

<b>(a) Create two NumPy arrays, one for Republicans and one for Democrats, that each contain the values above, in the order we provided them. Name these arrays `rep_trust` and `dem_trust`, respectively.</b> 

In [1]:
import numpy as np #imports numpy library, sets it as variable np
#creates numpy arrays of values (given above) and stores it in corresponding variable(rep_trust, dem_trust)
rep_trust = np.array([11,11,28,22,21]) 
dem_trust = np.array([17,26,15,15,4])

<b>(b) Write code to show that both arrays you created are of type array.</b> 

In [2]:
type(rep_trust) #calls built-in function type on rep_trust to find what datatype is stored in rep_trust

numpy.ndarray

In [3]:
type(dem_trust) #finds what datatype is stored in dem_trust

numpy.ndarray

<b>(c)  Make use of the built-in function `sum` to calculate the mean of your array of `rep_trust` prices and store the result as a variable called `rep_trust_mean`. Do all of this in one line of code.</b> 

In [4]:
rep_trust_mean = sum(rep_trust)/5 #sums the values in the array stored in rep_trust, divides them by number of values, 5

<b>(d)  Import the "mean" module from the statistics package (don't rename it as anything). Use it to calculate the mean of `dem_trust`. Store these results as a variable called `dem_trust_mean`. Do this in no more than two lines of code.</b>

In [5]:
from statistics import mean #imports mean function from statistics library
dem_trust_mean = mean(dem_trust) #finds mean of data stored in dem_trust + stores it in dem_trust_mean

<b>(e) Write one line of code to find out the data type of `rep_trust_mean`.</b> 

In [6]:
type(rep_trust_mean) #finds data type of what's stored in rep_trust_mean

numpy.float64

<b>(f) Write an expression that evaluates whether `rep_trust_mean` is greater than `dem_trust_mean`. Store this expression as a variable called `means_compare`.</b> 


In [7]:
means_compare = rep_trust_mean > dem_trust_mean #stores the comparison of if rep_trust_mean is great than dem_trust_mean to the variable means_compare
means_compare

True

**(g) What is the data type of `means_compare`? Why?** 

*Feel free to check the type in a cell above, but you don't need to turn in that work.*

It is a boolean. This is because when comparing the two means, I am checking to see if the rep_trust_mean is greater than dem_trust_mean with the > sign. This condition stored True in means_compare is the condition is true (rep_trust_mean > dem_trust_mean) or False if it's false, which is a boolean.

<b>(h) What is one conclusion about the real world can you draw based on your result in Question 2(f)?

On average, a higher percentage of republicans may have had trust in the government in Washington always or most of the time during the years of 2014-2019 (excluding 2018).

# Question 3: Your own study

Please answer the below questions based on the following scenario:

You decide to conduct your own study to understand the extent to which people living in New York City trust the national government in Washington, DC. To collect data, you go to Washington Square Park and ask people who are sitting there if they'd like to take a short, anonymous survey about trust in government. 
    
If they agree to participate, you ask them: "How confident are you that the government in DC will act in your best interest?" The possible answers are: Very confident, somewhat confident, not very confident, not at all confident, or prefer not to say.
    
You go around asking this question to strangers until you have 50 responses.

<b> (a) In this study, how are you conceptualizing "trust in government"?</b> 

I am conceptualizing trust in government as how confident people in NYC are that the government in DC will act in their best interest.

<b>(b) How are you operationalizing "trust in government"?</b> 


I am surveying people in Washington Square Park, asking them to answer Very confident, somewhat confident, not very confident, not at all confident, or prefer not to say to the question "How confident are you that the government in DC will act in your best interest?" I am collecting 50 anonymous surveys.

**(c) What is a possible source of random error, and why is it random?** 

A random error would be if I mishear some people's answers and record the wrong answer. It's random because it's a mistake on my part during the collection of data, and with a big enough sample, this one mistake would cancel out/not affect my data greatly.

**(d) What is a possible source of a systematic error, or selection bias, and in what direction would it bias your results?** 

Since I take my surveys at Washington Square Park, my results would come from people who are more likely to be at Washington Square Park, such as NYU students, people who live in or travel to Manhattan. Our results would not be representative of the entire NYC population, but rather biased towards certain groups of people.

<b> (e) Another data scientist points out that while you think you're measuring trust in government, the way you've worded your question means you might instead be picking up something more like optimism about government. This data scientist is raising the possibility of what kind of error? Briefly explain your answer.</b> 

They are raising the possibility of an error of validity. While my goal is to measure trust in government, I might be measuring optimism in government instead. My question phrasing might not be valid for measuring trust in government and instead, I might be picking up something else, which is an error of validity.

**(f) Suppose you import your results as a table into your Python program. You import the specific answers people gave, such as "very confident" or "somewhat confident" into your column named "responses".  As what type of data would this data be stored as by default in Python, and why?** 

This would be stored as a string because it is text/words written in quotation marks.

# Question 4: Importing data

Hooray! It's time to finally work with an actual dataset. Our data will come from the article linked in Question 1. The questions that follow refer to the data file "trust.csv", which is in this Homework 2 folder (where this assignment also is) on JupyterHub. This data reflects all the data in the second visualization ("Trust in govt higher among members of party that controls presidency"), the top five observations of which which we discussed in Question 2. 

**(a) Import the dataset as a pandas dataframe, name it `data`, and show the first five rows.** 

In [8]:
import pandas as pd #imports pandas library, sets it as pd
data = pd.read_csv('trust.csv') #imports dataset
data.head() #shows first 5 rows

date source  No Partisan Affiliation  Republican  Democrat
0  3/25/19    PEW                       22          21        14
1  12/4/17    PEW                       15          22        15
2  4/11/17    PEW                       12          28        15
3  10/4/15    PEW                       16          11        26
4  7/20/14    CNN                        8          11        17

**(b) How many observations and variables are there in this dataset? Write one line of code to find out, and then explain the answer in a Markdown cell below your code cell.** 

In [9]:
data.shape #shows (rows, columns)

(147, 5)

data.shape shows (rows, columns) of the dataset stored in the variable data. Rows are observations so there are 147 observations and columns are variables so there are 5 variables.

**(c) Whenever we import a new dataset, one of the first things we usually like to do is rename the variables so they are easy to refer to in our program. Rename the variables as follows: `No Partisan Affiliation` as `non_partisan`, `Republican` as `rep`, and `Democrat` as `dem`. Do so using just one command used one time.** 

In [10]:
data.rename(columns={'No Partisan Affiliation': 'non_partisan', 'Republican': 'rep', 'Democrat':'dem'}, inplace = True) #replaces the name of variables 'No Partisan Affiliation' to 'non_partisan', 'Republican' as 'rep', and 'Democrat' as 'dem'

**(d) What *variable* (not data) type is the variable `source`?** 

Categorical (specifically nominal)

**(e) Find the highest Republican level of trust in government by sorting the data so this observation is on top. What is this level, and in what year was it observed?** 

In [11]:
data.sort_values('rep', ascending = False, inplace = True) #sorts the table in data in place so that the variable rep shows values in descending order
data

date   source  non_partisan  rep  dem
146   12/1/58      NES            71   79   71
145  10/15/64      NES            74   73   80
54   10/25/01  CBS/NYT            50   70   47
55    10/6/01   GALLUP            53   68   52
119  11/30/86  CBS/NYT            54   63   37
..        ...      ...           ...  ...  ...
12   10/19/11  CBS/NYT            10    8   13
9      2/6/13  CBS/NYT            15    8   34
8     5/31/13  CBS/NYT            20    8   30
6    11/15/13  CBS/NYT            12    8   31
17   10/21/10  CBS/NYT            18    7   36

[147 rows x 5 columns]

This level is 79% and it was observed in 1958.

**(f) Find the highest Democratic level of trust in government by sorting the data so this observation is on top. What is this level, and in what year was it observed?** 

In [12]:
data.sort_values('dem', ascending = False, inplace = True) #sorts the table in data in place so that the variable dem shows values in descending order
data

date   source  non_partisan  rep  dem
145  10/15/64      NES            74   73   80
146   12/1/58      NES            71   79   71
144   12/1/66      NES            65   54   71
143  10/15/68      NES            56   60   66
43   10/15/04      NES            42   37   62
..        ...      ...           ...  ...  ...
2     4/11/17      PEW            12   28   15
1     12/4/17      PEW            15   22   15
0     3/25/19      PEW            22   21   14
12   10/19/11  CBS/NYT            10    8   13
31   10/13/08  CBS/NYT            28   19   12

[147 rows x 5 columns]

This level is 80% and it was observed in 1964.

**(g) Suppose we learn that all the polls conducted by "CBS/NYT" were actually only completed by CBS, not in partnership with the NYT (*New York Times*). Using just one command one time, rename all observations of "CBS/NYT" to just "CBS" in a way that the change will persist in your dataset (i.e., this change will hold as you continue your analyses).** 

In [13]:
data['source'].replace(['CBS/NYT'], 'CBS', inplace=True) #replaces 'CBS/NYT' to 'CBS' under the source variable in its place.

## Question 5: Descriptive statistics

For this question we will continue to use the above dataset to explore descriptive statistics.


<b> (a) Use the describe( ) function to generate descriptive statistics for the entire dataset. What variables are in the results, and why?</b> 

In [14]:
data.describe() #show descriptive stats for data stored in numerical variables

non_partisan         rep         dem
count    147.000000  147.000000  147.000000
mean      27.435374   34.265306   32.979592
std       12.523440   17.158139   10.939729
min        6.000000    7.000000   12.000000
25%       18.000000   20.000000   26.500000
50%       27.000000   31.000000   32.000000
75%       33.500000   50.000000   38.000000
max       74.000000   79.000000   80.000000

Political affiliations are in the results (non-partisan, republican, democrat) because they hold numerical data - the percentage of people of that party that trust the govt. describe() shows descriptive statistics for all numeric variables. Their variable type is numeric, and so we are able to calculate their descriptive stats (values such as mean, median). 

<b> (b) What is the interquartile range for `dem`, and what does this tell us, in substantive terms, about the variable?</b> 

In [15]:
from scipy.stats import iqr #imports iqr function from scipy.stats module
iqr(data['dem']) #finds iqr for dem

11.5

The interquartile range is 11.5. This tells us that the difference between the median of the 75th and 25th percentile of values, or the percentages of Democrats who trust the government at certain dates and from certain sources, is 11.5%. This means there is an scope of values of % of Democrats' trust in government in the middle 50% is within 11.5%.

<b> (c) What is the standard deviation for `dem`, and what does this tell us, in substantive terms, about the variable?</b> 

In [16]:
data['dem'].std() #finds the standard deviation for dem

10.939728645901612

The standard deviation is 10.939728645901612. This tells us that the values, or percentages of Democrats who trust the government at certain dates and from certain sources, are 10.939728645901612% spread out from the average percentage of Democrats who trust the government.

<b> (d) One of the arguments in this article is that democrats and republicans may have differing levels of trust in the government, and that it can vary depending on what party the president is. Create a subset of data that includes observations only from one four-year period that was under a republican administration (you might need to look up the dates!). Call this new dataframe `data_rep_admin`. In a Markdown cell, share which administration you've chosen.

In [17]:
data['date'] = pd.to_datetime(data['date']) #converts dates in variable data to datetime format (using pandas function).
mask = (data['date'] > '2017-1-20') & (data['date'] < '2021-1-20') #check if dates in the variable date are within range for the dates to look for (Trump's admin is between Jan 20, 2017-Jan 20, 2021)
data_rep_admin = data.loc[mask] #stores all the observations where the condition from above is true into data_rep_admin variable
data_rep_admin

date source  non_partisan  rep  dem
2 2017-04-11    PEW            12   28   15
1 2017-12-04    PEW            15   22   15
0 2019-03-25    PEW            22   21   14

Trump administration

<b>(e) Write code to find out: What was the mean level of republican trust in government during the administration you chose in 5(d) above? What was the mean level of democratic trust in government at that time? </b> 

In [18]:
mean(data_rep_admin['rep']) #finds mean level of republican trust in government during Trump admin

23.666666666666668

In [19]:
mean(data_rep_admin['dem']) #finds mean level of democratic trust in government at that time

14.666666666666666

<b> (f) Is your result consistent with the theory that republicans and democrats trust the government more when their own party holds the presidency? Why or why not?

No, this result does not tell us anything about our theory. The theory is that  republicans and democrats trust the government more when their own party holds the presidency, so to see if our results are consistent, we would have to gather data about republicans' and democrats' trust in government in a democratic presidency. Right now, our means are telling us about the trust in government in a republican presidency, but we also have to see if republicans have a lower mean during a democratic presidency and that democrats have a higher mean during a democratic presidency to say that our results are consistent with our theory.

**(g) Based on your answer to 5(f), what does this imply for the theory? Have you falisified it? Proved it? What's the verdict, and why?**

We have neither falsified it not proved it(which is not possible anyways). We don't have enough information to come to any type of conclusion with our theory. We also need observations from one four-year period that was under a republican administration so that we can find the means for the dem and rep trust in govt at that time and compare it to the means from the Trump administration's time to say anything about our theory, either falsify it or fail to disprove it.

## End of Homework

Congratulations on finishing the second homework! Don't forget to turn it in properly!